src/invidious/comments/youtube.cr

```diff
@@ -131,89 +133,148 @@ module Invidious::Comments
                   node_replies = node["replies"]["commentRepliesRenderer"]
                 end
 
-                if node["comment"]?
-                  node_comment = node["comment"]["commentRenderer"]
-                else
-                  node_comment = node["commentRenderer"]
-                end
-
-                content_html = node_comment["contentText"]?.try { |t| parse_content(t, id) } || ""
-                author = node_comment["authorText"]?.try &.["simpleText"]? || ""
-
-                json.field "verified", (node_comment["authorCommentBadge"]? != nil)
+                if node["commentViewModel"]?
+                  cvm = node.dig("commentViewModel", "commentViewModel")
+                  comment_key = cvm["commentKey"]
+                  toolbar_key = cvm["toolbarStateKey"]
+                  if mutations.size != 0
+                    comment_mutation = mutations.find { |i| i.dig?("payload", "commentEntityPayload", "key") == comment_key }
+                    toolbar_mutation = mutations.find { |i| i.dig?("entityKey") == toolbar_key }
+                    if !comment_mutation.nil? && !toolbar_mutation.nil?
+                      # todo parse styleRuns, commandRuns and attachmentRuns for comments
+                      html_content = HTML.escape(comment_mutation.dig("payload", "commentEntityPayload", "properties", "content", "content").as_s)
```

Comment:
Parsing the `commandRuns` is trivial, actually.
  
We already have a function that does that for descriptions here:
https://github.com/iv-org/invidious/blob/b673695aa2704b880562399ac78659ad23b7940d/src/invidious/videos/description.cr

This function should be hardened against XSS, though (it's not a concern in descriptions, as `<` and `>` are not allowed in there).

```diff
     cp = iter.next
     break if cp.is_a?(Iterator::Stop)
 
-    str << cp.chr
+    if cp == 0x26  # Ampersand (&)
+      str << "&amp;"
+    elsif cp == 0x27  # Single quote (')
+      str << "&#39;"
+    elsif cp == 0x22  # Double quote (")
+      str << "&quot;"
+    elsif cp == 0x3C  # Less-than (<)
+      str << "&lt;"
+    elsif cp == 0x3E  # Greater than (>)
+      str << "&gt;"
+    else
+      str << cp.chr
+    end
 
     # A codepoint from the SMP counts twice
     copied += 1 if cp > 0xFFFF
     copied += 1
```

Reply:
Would this not mess with the index included in the commandRuns?

Reply:
Not at all ^^

This function was specifically made to copy each code point individually while keeping track of the number of original characters this represents.

This is because the Crystal Stdlib doesn't allow UTF-16 Surrogates in an UTF-8 string (they are always decoded to their UTF-8 equivalent) but Youtube uses them everywhere.

Reply:
I was getting some weird formatting issues when trying to add parsing for styleRuns and attachmentRuns so I won't include it for now. This is the code I had, it would place links in the wrong spot in the description for some reason (it might be caused by overlapping runs?):
```crystal
def parse_runs(content : JSON::Any, video_id : String) : String
  text_content = content["content"].as_s
  return "" if text_content.empty?

  command_runs = content["commandRuns"]?.try &.as_a || [] of JSON::Any
  attachment_runs = content["attachmentRuns"]?.try &.as_a || [] of JSON::Any
  style_runs = content["styleRuns"]?.try &.as_a || [] of JSON::Any

  all_runs = (command_runs + attachment_runs + style_runs ).sort { |a, b| a["startIndex"].as_i <=> b["startIndex"].as_i }
  return text_content if all_runs.size == 0
  # Not everything is stored in UTF-8 on youtube's side. The SMP codepoints
  # (0x10000 and above) are encoded as UTF-16 surrogate pairs, which are
  # automatically decoded by the JSON parser. It means that we need to count
  # copied byte in a special manner, preventing the use of regular string copy.
  iter = text_content.each_codepoint
  index = 0
  return String.build do |str|
    all_runs.each do |command|
      cmd_start = command["startIndex"].as_i
      cmd_length = command["length"].as_i
      # Copy the text chunk between this command and the previous if needed.
      length = cmd_start - index
      index += copy_string(str, iter, length)
      # We need to copy the command's text using the iterator
      # and the special function defined above.
      cmd_content = String.build(cmd_length) do |str2|
        copy_string(str2, iter, cmd_length)
      end
      link = cmd_content
      if on_tap = command.dig?("onTap", "innertubeCommand")
        link = parse_link_endpoint(on_tap, cmd_content, video_id)
      elsif emoji_thumb = command.dig?("element", "type", "imageType", "image", "sources", 0)
        emoji_url = emoji_thumb["url"].as_s
        if emoji_url.starts_with?("https://lh3.googleusercontent.com/")
          link = String.build do |str|
            str << %(<img alt=") << cmd_content << "\" "
            str << %(src="/ggpht) << URI.parse(emoji_url).request_target << "\" "
            str << %(title=") << cmd_content << "\" "
            str << %(width=") << emoji_thumb["width"] << "\" "
            str << %(height=") << emoji_thumb["height"] << "\" "
            str << %(class="channel-emoji" />)
          end
        end
      end

      link = "<i>#{link}</i>" if command.dig?("italics")
      link = "<s>#{link}</s>" if command.dig?("strikethrough").try &.as_s == "LINE_STYLE_SINGLE"
      if weight_label = command.dig?("weightLabel").try &.as_s
        # Comments use MEDIUM for bold text and video descriptions use BOLD for bold text
        link = "<strong>#{link}</strong>" if weight_label == "FONT_WEIGHT_MEDIUM" || weight_label == "FONT_WEIGHT_BOLD"
      end
      str << link
      index += cmd_length
    end
    # Copy the end of the string (past the last command).
    remaining_length = content.size - index
    copy_string(str, iter, remaining_length) if remaining_length > 0
  end
end

```